### DISCUSSION TITLE
Can I set a pkl property value outside of an object declaration?

### DISCUSSION THREAD

**Discussion by ericzbeard** [apple/pkl#310] [Q&A] [answered]
Is there a syntax for doing this in pkl? The following fails:

```pkl
class Foo {
    Bar: String
}

f = Foo {
    Bar = "Baz"
}

f.Bar = "Something else"

output {
    value = f
}
```

```
–– Pkl Error ––
Mismatched input: `.`. Expected one of: `{`, `=`, `:`

9 | f.Bar = "Something else"
     ^
at fail (file:///Users/ezbeard/Source/pkl/fail.pkl, line 9)
```

I also can't set a property value within a class method.

**Comment by bioball** [comment, marked as answer]
No; Pkl values are immutable, and there are also no statements in Pkl.

You can express another object that is derived from `f` using object amending:

```groovy
class Foo {
    Bar: String
}

f = new Foo {
    Bar = "Baz"
}

g = (f) {
  Bar = "Something else"
}
```

**Reply by ericzbeard** [reply]
Thanks! This should be more obvious in the language reference. I missed the short note on immutability in the Concepts section and spent hours trying to figure out why I couldn't do this.

I would think that the flexibility offered by being able to mutate things would outweigh any potential downside, but I'm going to assume that's a design decision that won't change.

**Reply by bioball** [reply]
I've done some thought experiments around mutable scopes, something like...

```
foo = @{                   // made up syntax for a mutable scope
  bar = new Bar { baz = 15 }
  bar.baz += 15        // mutate `bar` within the scope
  return bar
  }@
```

The issue is that this would expand our language footprint by _quite_ a lot. It would add _statements_ to the grammar. Also, it would probably add features that would compete with already existing features (e.g. for loops vs. for generators), which would be a pain point for us. It would be a whole new paradigm (imperative, rather than declarative), which overall seems really hard to do right.
